rfd/0001-block-registry.md

```diff
@@ -0,0 +1,95 @@
+
+### Exposing TCMU to the agent container
+
+The Hocus install script NEEDS to install the following udev rule on the host OS:
+
+```
+# Hardlink hocus block devices using their serial number
+SUBSYSTEM=="block", ACTION=="add|change", ENV{ID_SCSI_SERIAL}=="hocusbd-[0-9A-Za-z]*", RUN+="/bin/sh -c 'mkdir -p /dev/hocus && ln $env{DEVNAME} /dev/hocus/$env{ID_SCSI_SERIAL}'"
+SUBSYSTEM=="block", ACTION=="remove", ENV{ID_SCSI_SERIAL}=="hocusbd-[0-9A-Za-z]*", RUN+="/bin/sh -c 'unlink /dev/hocus/$env{ID_SCSI_SERIAL}'"
+```
+
+This exposes block devices from the block registry in `/dev/hocus` on the host, the `/dev/hocus` folder is bind mounted into the Agent container. The block registry MUST consider that multiple agents might be running on the same computer and it does not have exclusivity over `/dev/hocus`. `/sys/kernel/config` and `/dev/uio0` will also be bind mounted in the container.
+
+### Block registry interface
+
+```js
+interface Image {
+  hash: string
+}
+
+interface Container {
+  hash: string
+}
+
+interface BlockRegistry {
+  getImages(): Image[]
+  getContainers(): Container[]
+  deleteContainer(container: Container): void
+  deleteImage(image: Image): void
+
+  // Loads an image from buildfs transferred via virtio-fs or downloaded from the internet using skopeo
+  loadImageFromDisk(path: path_to_oci_dump): Image
+  // Creates an RW layer on top of an image, creates an random hash
+  createContainer(image: Image): Container
+  // Converts the RW layer back into an RO layer, this deletes the container as overlaybd does not support cow snapshots
+  commitContainer(container: Container): Image
+
+  // Creates/Destroys an block device exposing the Image/Container. Images will be exposed as RO block devices, Containers as RW block devices.
+  expose(what: Image | Container): BlockDev
+  hide(what: Image | Container): void
+
+  /* FUTURE: P2P, perhaps expose an Docker registry api? */
+}
+```
+
+### Transferring images between buildfs and the block registry
+
+Use virtio-fs for it:
+
+1. Attach an empty agent directory to the buildfs vm using virtio-fs
+2. buildfs exports the image to the shared directory
+3. Agent loads the image into the block registy using loadImageFromDisk
+
+### Overlaybd block registry storage
```

Comment:
Please edit so it allows for arbitrary paths on disk for layers, containers, and images.

Reply:
done